File: android/core/src/main/kotlin/pes/core/Fold.kt
/**
 * Sample-state fold (spec §5.2). Mirrors `pes/core/fold.py`.
 *
 * foldSample reduces all events for one sample_id (across all devices and
 * files) to a single row, returned as a JsonObject shaped exactly like the
 * Python fold's dict (which is what `spec/fold_vectors.json` pins).
 * answered_at / latency_s come from the winning chain's root (original
 * submission); answer content comes from the chain's latest (effective) event.
 */
package pes.core

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/** (sourceFilePath, lineIndex, event) — any order. */
typealias EventTriple = Triple<String, Int, JsonObject>

/** Highest wins; `pending` / `scheduled` are derived, not event types. */
val STATUS_PRECEDENCE = listOf("retracted", "answered", "skipped", "expired", "suppressed", "unobserved")

/** Dedup key (§5.2 rule 0): (dev, sample, ev, t, n?, supersedes?). */
fun identityKey(ev: JsonObject): List<String?> {
    val key = mutableListOf(ev.optStr("dev"), ev.optStr("sample"), ev.optStr("ev"), ev.optStr("t"))
    if (ev.str("ev") == "snoozed") key.add((ev["n"] as? JsonPrimitive)?.content)
    if (ev.str("ev") == "answered") key.add(ev.optStr("supersedes"))
    return key
}

/**
 * Deduplicate by identity key; first by (file path, line) order wins.
 * Returns (kept events, warning count). Same key with a different payload is
 * a data error: first kept, warning recorded.
 */
fun dedupe(events: List<EventTriple>): Pair<List<JsonObject>, Int> {
    val ordered = events.sortedWith(compareBy({ it.first }, { it.second }))
    val seen = mutableMapOf<List<String?>, JsonObject>()
    val kept = mutableListOf<JsonObject>()
    var warnings = 0
    for ((_, _, ev) in ordered) {
        val key = identityKey(ev)
        val existing = seen[key]
        if (existing != null) {
            if (existing != ev) warnings += 1
            continue
        }
        seen[key] = ev
        kept.add(ev)
    }
    return Pair(kept, warnings)
}

/**
 * Resolve supersedes chains (§5.2 rule 2). Returns (winning chain root,
 * winning chain effective answer, other chains' effective answers).
 */
private fun answerChains(answered: List<JsonObject>): Triple<JsonObject?, JsonObject?, List<JsonObject>> {
    if (answered.isEmpty()) return Triple(null, null, emptyList())
    val existingTs = answered.map { it.str("t") }.toSet()
    val roots = answered.filter { it.optStr("supersedes")?.let { s -> s !in existingTs } ?: true }

    fun effective(root: JsonObject): JsonObject {
        var current = root
        val visited = mutableSetOf(System.identityHashCode(root))
        while (true) {
            val nexts = answered.filter {
                it.optStr("supersedes") == current.str("t") && System.identityHashCode(it) !in visited
            }
            if (nexts.isEmpty()) return current
            current = nexts.maxWith(compareBy({ it.str("t") }, { it.str("dev") }))
            visited.add(System.identityHashCode(current))
        }
    }

    val sortedRoots = roots.sortedWith(compareBy({ it.str("t") }, { it.str("dev") }))
    val winnerRoot = sortedRoots.first()
    return Triple(winnerRoot, effective(winnerRoot), sortedRoots.drop(1).map { effective(it) })
}

private fun firstConfigV(deduped: List<JsonObject>): Int? =
    deduped.filter { "config_v" in it }
        .minWithOrNull(compareBy({ it.str("t") }, { it.str("dev") }))
        ?.int("config_v")

/**
 * Fold one sample's events into a row. `expiryMinutes` is the effective
 * expiry at scheduled_utc, used only for `late`; null disables it.
 * Returns (row, warningCount).
 */
fun foldSample(events: List<EventTriple>, expiryMinutes: Int? = null): Pair<JsonObject, Int> {
    val (deduped, warnings) = dedupe(events)
    if (deduped.isEmpty()) {
        return Pair(JsonObject(mapOf("status" to JsonPrimitive("scheduled"))), warnings)
    }

    val sampleId = deduped.first().str("sample")
    val streamId = sampleId.substringBefore("|")
    val scheduledUtc = sampleId.substringAfter("|")
    val scheduledEpoch = parseUtc(scheduledUtc)

    val byType = deduped.groupBy { it.str("ev") }

    fun statusIgnoringRetraction(): String {
        for (status in STATUS_PRECEDENCE.drop(1)) {
            if (status in byType) return status
        }
        return if ("fired" in byType || "snoozed" in byType) "pending" else "scheduled"
    }

    val status: String
    val priorStatus: String?
    if ("retracted" in byType) {
        status = "retracted"
        priorStatus = statusIgnoringRetraction()
    } else {
        status = statusIgnoringRetraction()
        priorStatus = null
    }

    val (root, winner, duplicates) = answerChains(byType["answered"] ?: emptyList())
    val fired = byType["fired"] ?: emptyList()

    val latency = root?.let { parseUtc(it.str("t")) - scheduledEpoch }
    val late = if (winner != null && expiryMinutes != null) latency!! > expiryMinutes * 60L else false
    val suppressedReason = byType["suppressed"]
        ?.minWith(compareBy({ it.str("t") }, { it.str("dev") }))
        ?.optStr("reason")

    val row = linkedMapOf<String, kotlinx.serialization.json.JsonElement>(
        "sample" to JsonPrimitive(sampleId),
        "stream" to JsonPrimitive(streamId),
        "scheduled_utc" to JsonPrimitive(scheduledUtc),
        "status" to JsonPrimitive(status),
        "prior_status" to (priorStatus?.let { JsonPrimitive(it) } ?: JsonNull),
        "observed" to JsonPrimitive(fired.isNotEmpty()),
        "test" to JsonPrimitive(fired.any { it.bool("test", false) }),
        "snoozes" to JsonPrimitive(byType["snoozed"]?.size ?: 0),
        "fired_on" to JsonArray(fired.map { it.str("dev") }.distinct().sorted().map { JsonPrimitive(it) }),
        "answered_on" to (winner?.let { JsonPrimitive(it.str("dev")) } ?: JsonNull),
        "answered_at" to (root?.let { JsonPrimitive(it.str("t")) } ?: JsonNull),
        "latency_s" to (latency?.let { JsonPrimitive(it) } ?: JsonNull),
        "late" to JsonPrimitive(late),
        "partial" to JsonPrimitive(winner?.bool("partial", false) ?: false),
        "survey_id" to (winner?.let { JsonPrimitive(it.obj("survey").str("id")) } ?: JsonNull),
        "survey_version" to (winner?.let { JsonPrimitive(it.obj("survey").int("version")) } ?: JsonNull),
        "answers" to (winner?.optObj("answers") ?: JsonObject(emptyMap())),
        "loc" to (winner?.get("loc") ?: JsonNull),
        "suppressed_reason" to (suppressedReason?.let { JsonPrimitive(it) } ?: JsonNull),
        "duplicate_answers" to JsonPrimitive(duplicates.size),
        "config_version" to (firstConfigV(deduped)?.let { JsonPrimitive(it) } ?: JsonNull),
    )
    return Pair(JsonObject(row), warnings)
}

File: desktop/pes/core/fold.py
"""Sample-state fold (spec §5.2).

``fold_sample`` reduces all events for one sample_id (across all devices and
files) to a single row. It is a pure function: order-invariant and
duplication-invariant after the identity-key deduplication step, which is the
property the whole sync design rests on.
"""

from __future__ import annotations

from .timeutil import parse_utc

# Highest wins. `pending` / `scheduled` are derived, not event types.
STATUS_PRECEDENCE = ["retracted", "answered", "skipped", "expired", "suppressed", "unobserved"]


def identity_key(ev: dict) -> tuple:
    """Dedup key (§5.2 rule 0): (dev, sample, ev, t, n?, supersedes?)."""
    key = [ev["dev"], ev.get("sample"), ev["ev"], ev["t"]]
    if ev["ev"] == "snoozed":
        key.append(ev.get("n"))
    if ev["ev"] == "answered":
        key.append(ev.get("supersedes"))
    return tuple(key)


def dedupe(events: list[tuple[str, int, dict]]) -> tuple[list[dict], list[str]]:
    """Deduplicate by identity key; first by (file path, line) order wins.

    Returns (kept events, warnings). Same key with a different payload is a
    data error: first kept, warning recorded.
    """
    ordered = sorted(events, key=lambda e: (e[0], e[1]))
    seen: dict[tuple, dict] = {}
    kept: list[dict] = []
    warnings: list[str] = []
    for path, _line, ev in ordered:
        key = identity_key(ev)
        if key in seen:
            if seen[key] != ev:
                warnings.append(f"conflicting duplicate for {key} in {path}")
            continue
        seen[key] = ev
        kept.append(ev)
    return kept, warnings


def _answer_chains(answered: list[dict]) -> tuple[dict | None, dict | None, list[dict]]:
    """Resolve supersedes chains (§5.2 rule 2).

    Returns (winning chain root, winning chain effective answer, other
    chains' effective answers). Chains are keyed by the `t` of the superseded
    event; an event whose `supersedes` target is absent starts its own chain.
    The effective answer of a chain is its latest event (ties broken by dev,
    ascending, last wins); the winning chain is the one whose root has the
    earliest `t` (tie: dev).

    The root matters separately from the effective answer: answered_at and
    latency are measured from the original submission (the root), so a later
    edit never changes them; answer content comes from the effective event.
    """
    if not answered:
        return None, None, []
    existing_ts = {ev["t"] for ev in answered}
    roots = [ev for ev in answered if not ev.get("supersedes") or ev["supersedes"] not in existing_ts]
    if not roots:
        # Corrupt data: every event supersedes another (a cycle). The fold
        # must stay total — treat all events as chain roots rather than crash
        # every future refold of this sample.
        roots = list(answered)

    def effective(root: dict) -> dict:
        current = root
        visited = {id(root)}
        while True:
            nexts = [
                ev
                for ev in answered
                if ev.get("supersedes") == current["t"] and id(ev) not in visited
            ]
            if not nexts:
                return current
            current = max(nexts, key=lambda ev: (ev["t"], ev["dev"]))
            visited.add(id(current))

    roots.sort(key=lambda ev: (ev["t"], ev["dev"]))
    winner_root, *other_roots = roots
    return winner_root, effective(winner_root), [effective(r) for r in other_roots]


def fold_sample(
    events: list[tuple[str, int, dict]], expiry_minutes: int | None = None
) -> dict:
    """Fold one sample's events into a row.

    ``events``: (source_file_path, line_index, event) triples, any order.
    ``expiry_minutes``: effective expiry at scheduled_utc (stream override ->
    global default), used only for the ``late`` computation; None disables it.
    """
    deduped, warnings = dedupe(events)
    if not deduped:
        return {"status": "scheduled", "warnings": warnings}

    sample_id = deduped[0]["sample"]
    stream_id, scheduled_utc = sample_id.split("|", 1)
    scheduled_epoch = parse_utc(scheduled_utc)

    by_type: dict[str, list[dict]] = {}
    for ev in deduped:
        by_type.setdefault(ev["ev"], []).append(ev)

    def status_ignoring_retraction() -> str:
        for status in STATUS_PRECEDENCE[1:]:
            if status in by_type:
                return status
        if "fired" in by_type or "snoozed" in by_type:
            return "pending"
        return "scheduled"

    if "retracted" in by_type:
        status = "retracted"
        prior_status = status_ignoring_retraction()
    else:
        status = status_ignoring_retraction()
        prior_status = None

    root, winner, duplicates = _answer_chains(by_type.get("answered", []))

    fired = by_type.get("fired", [])
    row = {
        "sample": sample_id,
        "stream": stream_id,
        "scheduled_utc": scheduled_utc,
        "status": status,
        "prior_status": prior_status,
        "observed": bool(fired),
        "test": any(ev.get("test", False) for ev in fired),
        "snoozes": len(by_type.get("snoozed", [])),
        "fired_on": sorted({ev["dev"] for ev in fired}),
        "answered_on": winner["dev"] if winner else None,
        "answered_at": root["t"] if root else None,
        "latency_s": (parse_utc(root["t"]) - scheduled_epoch) if root else None,
        "late": False,
        "partial": bool(winner.get("partial", False)) if winner else False,
        "survey_id": winner["survey"]["id"] if winner else None,
        "survey_version": winner["survey"]["version"] if winner else None,
        "answers": winner.get("answers", {}) if winner else {},
        "loc": winner.get("loc") if winner else None,
        "suppressed_reason": (
            min(by_type["suppressed"], key=lambda ev: (ev["t"], ev["dev"])).get("reason")
            if "suppressed" in by_type
            else None
        ),
        "duplicate_answers": len(duplicates),
        "config_version": _first_config_v(deduped),
        "warnings": warnings,
    }
    if winner and expiry_minutes is not None:
        row["late"] = row["latency_s"] > expiry_minutes * 60
    return row


def _first_config_v(deduped: list[dict]) -> int | None:
    carrying = [ev for ev in deduped if "config_v" in ev]
    if not carrying:
        return None
    return min(carrying, key=lambda ev: (ev["t"], ev["dev"]))["config_v"]
